0012-common-ledger-api/0012-common-ledger-api.md

```diff
@@ -0,0 +1,1181 @@
+
+
+### Account Resource
+[Account resource]: #account-resource
+
+An account resource contains the fields in the following table. Some fields are _Server-provided_, meaning they cannot be set directly by clients. Fields that are "Optional" or "Server-provided" in the following table may be omitted by clients submitting objects to the API, but those fields are not optional to implement. Fields of nested objects are indicated with a dot (`.`) character; no fields contain a dot literal.
+
+| Name                      | Type    | Description                            |
+|:--------------------------|:--------|:---------------------------------------|
+| `name`                    | String  | Name of the account. A ledger MAY require this to be unique. MUST match the regular expression `^[a-zA-Z0-9._~-]{1,256}$`. |
+| `fingerprint`             | String  | _(Optional)_ A fingerprint of the account's client certificate. This field MUST be available if and only if the ledger supports client certificate authentication. MUST match the regular expression `^[0-9A-Fa-f]{2}(:[0-9A-Fa-f]{2}){19,127}$`. |
+| `ledger`                  | [URL][] | _(Optional)_ The path to the ledger containing this account. MUST be an HTTP(S) URL where the client can [Get Ledger Metadata][]. |
+| `minimum_allowed_balance` | String  | _(Optional)_ The minimum balance permitted on this account. The special value `"-infinity"` indicates no minimum balance. This is a string so that no precision is lost in JSON encoding/decoding. The default value SHOULD be `"0"`. |
+| `id`                      | [URL][] | _(Server-provided)_ The primary, unique identifier for this account. MUST be an HTTP(S) [URL][] where the client can [get this resource][Get Account]. |
+| `balance`                 | String  | _(Server-provided)_ Balance as decimal [amount][]. Defaults to `"0"`. This can be negative, if the account's `minimum_allowed_balance` allows it. |
+
+A ledger MAY define additional fields for the account resource. For example, `is_admin` could be a boolean that, if true, grants administrator-level authorization when authenticated as the account owner.
+
+
+### Message Resource
+[Message resource]: #message-resource
+
+Messages are sent through the ledger's [Send Message][] method and received in a [WebSocket][] subscription. All fields of the message are required:
+
+| Field    | Value   | Description                                             |
+|:---------|:--------|:--------------------------------------------------------|
+| `ledger` | [URL][] | The base [URL][] of this ledger. MUST be an HTTP(S) URL where the client can [Get Ledger Metadata][]. |
```

Comment:
I think it's fairly common in a REST API to provide a resource property that is the base URL for the resource, especially in the case of resources that may be sent over something like websockets where the transport should be decoupled from the message content.

In fact if we're being strict about principles then the hypermedia principle of REST says we MUST have that property. That said, REST API design always becomes a compromise between convention and configuration. If we can assert that for this API the ledger URL for an account is always `../../` relative to the account URL (`https://ledger.example/accounts/alice`) then perhaps that's enough.

Finally, it seems very flaky to me to connect to `ws://ledger.example` and then get notifications with `from: alice` and `to: bob` as properties that I must figure out, based on the URL I connected to originally, actually resolve to `from: https://ledger.example/accounts/alice` and `to: https://ledger.example/accounts/bob`.

Perhaps the answer is to make it optional for the REST API calls and always present in websocket messages? (Not sure what behaviour you'd want to mandate though if the ledger URL property doesn't correspond to the URL of the resource itself, i.e. share a common base)